content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+of "regular structure" is given in the implementation detail section. 
+- For each loop that can be unrolled, leverage the information from constant propagation and reaching definition passes to calculate the tripcount. 
+- If the tripcount is computable, decide whether we want to unroll the loop or not using a simple criterion: the total 
+number of instructions in the loop (tripcount * number of instructions in all loop basic blocks). Currently we unroll loops
+with less than 1k instructions. 
+- Unroll the loop by first duplicating all basic blocks inside the loop and then removing redundant control flow instructions. 
+
+#### Dynamic Profiling
+
+The key idea of dynamic profiling is to estimate the loop status by simulating the loop execution. Since the loop trip count may depend on different conditions, it will hard to infer the loop bound without the information from actual execution environment. For example, induction variable being udpated indirectly can make the unrolling problem complicated. An intuitive way of simplifying the problem is to proflie the loops before unrolling. Here we extend the data flow analysis pass to support dynamic profiling, i.e. we maintain a status table to track the update history for each variables and . By refering to the table and locate the update history of the back-edge related instructions and basic blocks, we can easily infer various information needed, e.g. trip count, induction variable as well as loop stride. Here to simplify the analysis, we assume all loops to be unrolled are natural loops, namely the program can only enter the loop from the header. Under the assumption that the loop bound is invariant in multiple runs, the basic blocks in the loop body will be repliacted for trip bound times. Back edges (i.e. jmps) and condition check statements will be removed form originla bril program.
+
+### Implementation details
+
+#### External Packages
+In this project, we use two external packages. The first one is the array extension of Bril, whose source code can be found [here](https://github.com/Checkmate50/bril/tree/arrays). The second one is the C code generation for Bril. The source code can be found [here](https://github.com/xu3kev/bril2c).
+
+#### Static Analysis
+This section details how we implement the static loop unrolling pass. We first state our definition of "regular loops" that
+can be unrolled by our pass. A "regular loop" should have the following properties other than the ones mentioned in the 
+input specification section:
+- The control flow can only leave the loop from either of the two blocks connected by the back-edge. 
+- The loop can not contain any sub-loop. 
+
+In this section, without special mentioning, we will use *entry* to denote the entrance block of the loop (destination of 
+the backedge), and use *exit* to denote the exit block of the loop (source of the backedge). The implementation details of 
```

Comment:
Consider staying consistent about whether you use "backedge," "back edge," or "back-edge."